# ¡Hola, Gabriel!  

Mi nombre es Carlos Ortiz, soy code reviewer de Practicum y voy a revisar el proyecto que acabas de desarrollar.

Cuando vea un error la primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión. 

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

¡Empecemos!

# Proyecto de Preporcesamiento de Datos.

Por: *Fernando Gabriel Ponce Lara.*

## Tabla de Contenidos <a class="anchor" id="tablacontenido"></a>

+ [Introducción](#Introducción)
+ [Planteamiento del Problema](#planteamiento)
+ [01. Exploración de datos](#primertitulo)
    + [Descripción de los datos](#primertitulo_sub1)
    + [Conclusión de la Sección](#exploracionConclusion)
+ [02. Transformación de datos](#transformacion)
+ [03. Trabajar con valores ausentes](#ausentes)
+ [04. Clasificación de datos](#clasificar)
+ [05. Comprobación de las hipótesis](#comprobacion)
+ [06. Conclusión General](#conclusiongeneral)



## Análisis del riesgo de incumplimiento de los prestatarios

El proyecto consiste en preparar un informe para la división de préstamos de un banco. Deberás averiguar si el estado civil y el número de hijos de un cliente tienen un impacto en el incumplimiento de pago de un préstamo. El banco ya tiene algunos datos sobre la solvencia crediticia de los clientes.

Este informe se tendrá en cuenta al crear una **puntuación de crédito** para un cliente potencial. La **puntuación de crédito** se utiliza para evaluar la capacidad de un prestatario potencial para pagar su préstamo.


##  Propósito del proyecto

El propósito del proyecto es analizar los datos de los clientes de la división de préstamos de un banco para determinar si el estado civil y el número de hijos tienen un impacto en el incumplimiento de pago de un préstamo. El informe resultante se utilizará para crear una puntuación de crédito para los clientes potenciales, lo que ayudará al banco a evaluar la capacidad de un prestatario potencial para pagar su préstamo.

Para llevar a cabo este proyecto, se analizarán los datos existentes de los clientes, incluyendo información como el número de hijos, la experiencia laboral, la edad, la educación, el estado civil, el género, el tipo de empleo, la deuda, el ingreso mensual y el propósito del préstamo. Se buscará patrones y relaciones entre estos datos y el incumplimiento de pago para determinar si el estado civil y el número de hijos son factores que influyen en el riesgo de incumplimiento.

El informe final se presentará al banco y se utilizará para mejorar la puntuación de crédito de los clientes potenciales, lo que ayudará al banco a tomar decisiones más informadas sobre el otorgamiento de préstamos y a reducir el riesgo de incumplimiento.


<div class="alert alert-block alert-info" >

<b>Nota estudiante:</b> <br>
    
    No eliminare lo que esta en corchetes por favor asumir que lo hice. no puedo trabajarlo  de forma local.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Hola, Gabriel. Dado que es una instrucción importante en el proyecto porque el objetivo de los mismos es que puedas utilizarlo como parte de tu portafolio en el futuro, quisiera preguntar ¿a qué haces referencia con que no puedes trabajarlo de manera local? Eliminar los comentarios solo requiere de que hagas doble clic en la celda y elimines lo que está escrito. Esto lo puedes lograr trabajando en jupyter, en la plataforma o en colab.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
En esta sección debes incluir una introducción al proyecto, los objetivos y una tabla de contenidos. Esto facilitará la navegación en el mismo. 
</div>

## Introducción 

Se busca analizar los datos de los clientes de la división de préstamos de un banco con el objetivo de determinar si existen relaciones entre el incumplimiento de pago de un préstamo y ciertos factores como el número de hijos, el estado civil y el nivel de ingresos. El informe resultante se utilizará para mejorar la puntuación de crédito de los clientes potenciales, lo que ayudará al banco a evaluar mejor la capacidad de pago de un prestatario potencial y tomar decisiones más informadas sobre el otorgamiento de préstamos.
Para esto se necesita un proceso de análisis de datos para comprobar las hipótesis que resuelven la problemática.

## Planteamiento del Problema  <a class="anchor" id="planteamiento"></a>

El problema planteado en este proyecto es determinar si el estado civil y el número de hijos de un cliente tienen un impacto en el incumplimiento de pago de un préstamo. El banco busca mejorar su proceso de evaluación de riesgos al momento de otorgar un préstamo, para esto se busca entender si estos factores son relevantes para determinar la capacidad de pago de un cliente y así poder incluirlos en la puntuación de crédito. El análisis de los datos existentes permitirá entender si estos factores son relevantes para el incumplimiento de pago y ayudará al banco a tomar decisiones más informadas al momento de otorgar un préstamo.

## Hipótesis a probar

Concretamente se plantean tres hipótesis a probar:

    1.Existe una relación entre el número de hijos de un cliente y el incumplimiento de pago de un préstamo.
    2.Existe una relación entre el estado civil de las personas casadas y el incumplimiento de pago de un préstamo.
    3.Existe una relación entre el nivel de ingresos de un cliente y el incumplimiento de pago de un préstamo.
    4.Afecta el propósito del crédito a la tasa de incumplimiento

## Abre el archivo de datos y mira la información general. 

[Empieza con la importación de las librerías y la carga de los datos. Es posible que te des cuenta de que necesitas librerías adicionales a medida que avanzas, lo cual es totalmente normal, solo asegúrate de actualizar esta sección cuando lo hagas.]

Empezamos abriendo e importando las librerías y el ‘dataset’ con el que trabajaremos en este proyecto.

In [143]:
# Cargar todas las librerías
import pandas as pd
import numpy as np
import math


In [144]:
# Carga los datos

try:
    df = pd.read_csv('credit_scoring_eng.csv')
except:
    df = pd.read_csv('/datasets/credit_scoring_eng.csv') 

# 01. Exploración de datos <a class="anchor" id="primertitulo"></a>

 **Descripción de los datos** <a class="anchor" id="primertitulo_sub1"> </a>

- `children` - el número de hijos en la familia
- `days_employed` - experiencia laboral en días
- `dob_years` - la edad del cliente en años
- `education` - la educación del cliente
- `education_id` - identificador de educación
- `family_status` - estado civil
- `family_status_id` - identificador de estado civil
- `gender` - género del cliente
- `income_type` - tipo de empleo
- `debt` - ¿había alguna deuda en el pago de un préstamo?
- `total_income` - ingreso mensual
- `purpose` - el propósito de obtener un préstamo



Ahora veremos la columnas y 10 primeras filas de nuestro `dataset`.

In [145]:
# Vamos a ver cuántas filas y columnas tiene nuestro conjunto de datos

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [146]:
# vamos a mostrar las primeras filas N

df.head(5)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding


#### Observaciones:

- Se puede concluir que el data frame cuenta con valores nulos, por lo que se debe considerar dar un tratamiento a los datos.

-  El ‘dataframe’ tiene en total 21525 entradas, pero solo dos columnas no tiene todos los valores completos en este caso las columna: `days_employed` y `total_income`.
- Por lo tanto, hay que enfocarse en esas 2 columnas especialmente.


In [147]:
# Veamos la tabla filtrada con valores ausentes de la primera columna donde faltan datos


dfl1=df.loc[df['days_employed'].isna(),:]
##tamnio tabla filtrada con valores vacios de `days_employed`
print('Primer Filtro \'days_employed\':' ,str('len(dfl1)') , len(dfl1))


### aplicacmos el filtro para vacios de `total_income`
dfl1l2=dfl1.loc[df['total_income'].isna(),:]

##tamnio tabla filtrada con valores vacios de `days_employed`
print('Doble Filtro \'total_income\':' ,str('len(dfl1l2)') , len(dfl1l2))

## valores ausentes e dataframe
print(str("dfl1l2['days_employed']") ,dfl1l2['days_employed'].isna().sum())


print('\ntabla filtrada: \n' )
dfl1l2.head()

Primer Filtro 'days_employed': len(dfl1) 2174
Doble Filtro 'total_income': len(dfl1l2) 2174
dfl1l2['days_employed'] 2174

tabla filtrada: 



children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding


**Observaciones:**

- Se puede ver que las columnas con valores ausente son:
    + `days_employed` con 2174 valores vacíos en dicha columna.
    + `total_income` con 2174 valores vacíos en dicha columna.

- Para ver si los valores ausentes son simétricos , se comprobó aplicando los 2 filtros y mantienen el mismo número de valores vacíos. Por lo tanto se Concluye que son simétricos.


In [148]:
# Apliquemos múltiples condiciones para filtrar datos y veamos el número de filas en la tabla filtrada.

##el filtro se aplico en las celdas anteriores
print("Numero de filas de la tabla filtrada (con valores vacios) : ",len(dfl1l2) )

### porcentaje de valores vacios.
porcentajeVACIO=len(dfl1l2)/len(df)
print(f'porcentaje de valores vacios: {porcentajeVACIO:.2%}')




Numero de filas de la tabla filtrada (con valores vacios) :  2174
porcentaje de valores vacios: 10.10%


#### Observaciones :


+ Se concluye que hay valores ausentes , no son iguales que la tabla original, Por lo tanto hay que dar un tratamiento a esas columnas.
+ Aproximadamente el 10% de los valores poseen valores ausentes en el campo de la columna `days_employed`
+ Por lo tanto hay que continuar con los pasos siguientes de completado de valores ausentes.

In [149]:
# Vamos a investigar a los clientes que no tienen datos sobre la característica identificada y la columna con los valores ausentes

dfl1l2.head(5)

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding

In [150]:
# Comprobación de la distribución

## imprimimos la edad
dfl1.describe()

children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

**Observaciones**:
+ Se puede ver que hay un patrón muy fuerte entre las columnas ,donde `days_employed` esta vacío  también lo está la columna `total_income`. Es posible que en el Registro de Datos o durante el procesamiento de estos ocurra un error que los haga aparecer como vacíos.
+ Se concluye que los valores ausentes no siguen un patrón es especifico por lo que debe de continuar con los siguientes pasos. 
+ Posiblemente los valores ausentes se dieron al ingreso de los datos.


In [151]:
# Comprobando la distribución en el conjunto de datos entero

df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

**Conclusión intermedia `Estudiante:`** 

+ Al comparar la distribución de datos del conjunto entero del DataFrame con valores vacíos, se puede apreciar que:
    - Los valores de las columnas se parecen mucho, excepto en la columnas donde hay valores vacíos.
+ Al apreciar la distribución de ambas se nota que no hay mucha diferencia ni varianza en los valores de las columnas no vacías.
+ No son aleatorios los valores vacíos, al parecer. Porque se concluyó que la misma cantidad de valores vacíos de la columna `days_employed	` es igual al de la columna `total_income`.



In [152]:
# Comprueba otras razones y patrones que podrían llevar a valores ausentes

##tabla original
print('tabla original:\n',df.describe())

## tabla filtrada
print('tabla filtrada:\n')
dfl1l2.describe()

tabla original:
            children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%

children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

**Conclusión intermedia**


+ Se nota que los valores no pueden ser 100 % accidentales , Existe un patrón.
+ Es la misma cantidad de valores ausente que la columna `total_income`
+ La misma cantidad de valores ausentes en dichas columnas `total_income` y `days_employed` por lo tanto que si se tratan los mismos estos completarían el DataSet.


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
En este punto sí puedes concluir que son simétricos.
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <br>
    Listo. 
    <a class="tocSkip"></a>
</div>



**Conclusión de la Sección** <a class="anchor" id="exploracionConclusion"></a>

   __Tratamiento de Datos__:

1. Se concluye que no hay un patrón en específico para este caso , porque al comparar los valores filtrados con la tabla original , se encontró que no difieren demasiado utilizando `df.describe()`
2. Abordaje de Valores Ausentes
    - se deben de sustituir los valores numéricos por un valor representativo.
    - o agruparlos por un valor categórico.
3. Transformación de datos.
    - se eliminará los duplicados.
    - se cambiará los tipos de ser necesario.
    - se rellenará los valores ausentes

    
 

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Es necesario que, para concluir si hay aleatoriedad o no en la existencia de los datos ausentes, compruebes las distribuciones por separado de algunas variables con y sin datos ausentes. Esto lo puedes hacer con el método `value_counts` pasándole el argumento `normalize=True`. Si los porcentajes difieren significativamente en alguna variable, se puede concluir que la razón depende de esa variable. Si no hay diferencias para ninguna, sí se puede concluir que son aleatorios.
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <br>
    Listo. Se agrego el `value_counts` , lo que justifica la Aletoriedad.
    <a class="tocSkip"></a>
</div>


<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Debemos trabajar un poco en esta sección:
    
1. Debes crear dos dataframes: uno con datos faltantes y otro sin datos faltantes.
    
2. Piensa en algunas columnas que pueden tener relación con los datos faltantes. Te puedo sugerir `income_type`, `family_status`, `debt` y `gender`.
    
3. Para cada dataframe debes utilizar el método `value_counts` pasándole el argumento `normalize=True` para cada columna. Por ejemplo, si se quiere analizar `income_type`, podemos escribir:
    
```python
   df['income_type'].value_counts(normalize=True)
   df_not_na['income_type'].value_counts(normalize=True)
```
    
Si existe alguna diferencia significativa entre estos dos resultados, se puede concluir que los datos faltantes tienen alguna relación con la columna en cuestión. Recuerda aplicarlo a cada columna y concluir sobre lo encontrado.
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante #2</b> <br>
    Listo. Se elimino la celda incorrecta y se hizo una comparacion entre solo valores nulos y no nulos. agrupados por `income_type`. Se lo hace a continuacion:
    <a class="tocSkip"></a>
</div>


In [153]:
# Crear dataframe con valores ausentes en la columna 'days_employed' y 'total_income'
df_solo_nulos_days_employed= df[df['days_employed'].isnull()]

# Crear dataframe sin valores ausentes en la columna 'days_employed' y 'total_income'
df_sin_nulos_days_employed= df.dropna(subset=['days_employed'])

# Analizar la distribución de income_type en ambos dataframes
print("Distribución de income_type en dataframe con valores ausentes de days_employed: \n", df_solo_nulos_days_employed['income_type'].value_counts(normalize=True),"\n")
print("Distribución de income_type en dataframe sin valores ausentes de days_employed: \n", df_sin_nulos_days_employed['income_type'].value_counts(normalize=True))


Distribución de income_type en dataframe con valores ausentes de days_employed: 
 employee         0.508280
business         0.233671
retiree          0.189972
civil servant    0.067617
entrepreneur     0.000460
Name: income_type, dtype: float64 

Distribución de income_type en dataframe sin valores ausentes de days_employed: 
 employee                       0.517493
business                       0.236525
retiree                        0.177924
civil servant                  0.067800
unemployed                     0.000103
student                        0.000052
entrepreneur                   0.000052
paternity / maternity leave    0.000052
Name: income_type, dtype: float64


In [154]:
# Crear dataframe con valores ausentes en la columna 'days_employed' y 'total_income'
df_solo_nulos_total_income= df[df['total_income'].isnull()]

# Crear dataframe sin valores ausentes en la columna 'days_employed' y 'total_income'
df_sin_nulos_total_income= df.dropna(subset=['total_income'])

# Analizar la distribución de income_type en ambos dataframes
print("Distribución de income_type en dataframe con valores ausentes de total_income: \n", df_solo_nulos_days_employed['income_type'].value_counts(normalize=True),"\n")
print("Distribución de income_type en dataframe sin valores ausentes de total_income: \n", df_sin_nulos_days_employed['income_type'].value_counts(normalize=True))


Distribución de income_type en dataframe con valores ausentes de total_income: 
 employee         0.508280
business         0.233671
retiree          0.189972
civil servant    0.067617
entrepreneur     0.000460
Name: income_type, dtype: float64 

Distribución de income_type en dataframe sin valores ausentes de total_income: 
 employee                       0.517493
business                       0.236525
retiree                        0.177924
civil servant                  0.067800
unemployed                     0.000103
student                        0.000052
entrepreneur                   0.000052
paternity / maternity leave    0.000052
Name: income_type, dtype: float64


Se sabe que la distribución con valores nulos y no nulos son aproximadamente simétricos , por lo que se concluye que existe algún tipo de simetría.

- [__Click para ir a la Tabla de Contenidos__](#tablacontenido)

# 02 Transformación de datos <a class="anchor" id="transformacion"></a>

En esta etapa vamos a Dar un tratamiento a Datos que puedan interferir con la calidad de los mismos durante su análisis.

In [155]:
# Veamos todos los valores en la columna de educación para verificar si será necesario corregir la ortografía y qué habrá que corregir exactamente

df.info()
df['education'].describe()
print(df['education'].unique())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME C

In [156]:
# Arregla los registros si es necesario
df['education'] = df['education'].str.lower()

In [157]:
# Comprobar todos los valores en la columna para asegurarnos de que los hayamos corregido

print(df['education'].unique())

["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo.
</div>

Datos de la columna `children`:

In [158]:
# Veamos la distribución de los valores en la columna `children`
df['children'].describe()

count    21525.000000
mean         0.538908
std          1.381587
min         -1.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         20.000000
Name: children, dtype: float64

Vamos a ver si esta Columna Resulta problemática por lo que vamos a ver cuantos valores erróneos puede Presentar.

In [159]:
# [arregla los datos según tu decisión]



##vemos que hay valores de -1 
dfl2=df.loc[df['children']<0 ,:]
print('valores menores que 0: \n',len(dfl2))

###valores igual a 20
dfl2_=df.loc[df['children']==20,:]
print('valores iguales a 20: \n',len(dfl2_))


###numero de valores erronesos
numero_erroneo_children=len(dfl2)+len(dfl2_)

##vemos que mayores que 9  
#dfl3=df.loc[df['children']>=9 ,:]
#print('valores mayores o iguales a 9: \n',len(dfl3))

###correccion
df['children']=abs(df['children']) ##aplicar el valor absoluto



### para children se creara una funcion para alterar unicamente esa columna

"""
def arreglar_children_igual_de_20(hijos):
    hijos_retorno=hijos
    if hijos == 20:
        hijos_retorno=2
        return hijos_retorno

    return hijos_retorno
"""

#df['children']= df['children'].apply(arreglar_children_igual_de_20)
df['children'] = df['children'].replace(20, 2)

valores menores que 0: 
 47
valores iguales a 20: 
 76


<div class="alert alert-block alert-warning">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Esto lo puedes hacer más fácil con el método `replace`.
</div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante #2</b> <br>
    Listo. se comento el codigo.
    <a class="tocSkip"></a>
</div>


+ Comprobación si hay valores erróneos después de la corrección.

In [160]:
df['children'].describe()

##vemos que hay valores de -1 
dfl2=df.loc[df['children']<0 ,:]
print('valores menores que 0: \n',len(dfl2))

###valores igual a 20 comprobacion
dfl2_=df.loc[df['children']==20,:]
print('valores iguales a 20: \n',len(dfl2_))


numero_erroneo_children=len(dfl2)+len(dfl2_)

propocion_erroneo_children=numero_erroneo_children/len(df)
print(f'Porcentaje de Valores Errorneos de la Columna children: {propocion_erroneo_children:.2%}')

valores menores que 0: 
 0
valores iguales a 20: 
 0
Porcentaje de Valores Errorneos de la Columna children: 0.00%


In [161]:
# Comprobar la columna `children` de nuevo para asegurarnos de que todo está arreglado

##vemos que hay valores de -1 
dfl2=df.loc[df['children']<0 ,:]
print('valores menores que 0: \n',len(dfl2))

###valores igual a 20
dfl2_=df.loc[df['children']==20,:]
print('valores iguales a 20: \n',len(dfl2_))

df['children'].describe()

valores menores que 0: 
 0
valores iguales a 20: 
 0


count    21525.000000
mean         0.479721
std          0.755528
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          5.000000
Name: children, dtype: float64

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo.
</div>

In [162]:
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [163]:
df['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)


Posiblemente la Columna `days_employed` puede presentar valores negativos por lo que hay que cambiarlos.
También Es posible que tenga Valores nulos pero eso lo trataremos más adelante.


In [164]:
# Encuentra datos problemáticos en `days_employed`, si existen, y calcula el porcentaje
print(df['days_employed'].describe(),'\n')

dfl4=df.loc[df['days_employed']<=0 ,:]

percentaje_days_employed=len(dfl4)/len(df)

print(f'porcentaje de valore negativos {percentaje_days_employed:.2%}')

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64 

porcentaje de valore negativos 73.90%


__Columna `days_employed`__

Podemos observar que los valores menores que 0 o negativos cubren el 73% de los datos por lo que es demasiado Grande, En este caso , Hay que avisar al programado porque hay tantos valores de esta característica.

+ En este caso Vamos a convertir los valores negativos de esta columna en valores positivos.

También hay que considerar que podemos ver valores atípicos. Muy grandes.
+ En este caso aplicaremos el filtro donde si el número de días se divide para 365 y esto resulta mayor que 50 entonces es un valor atípico.


In [165]:
# Aborda los valores problemáticos, si existen.


### en este caso usaremos el valor absoluto para corregir los negativos, suponiendo que solo  el signo es la falla
print(df['days_employed'])



0         -8437.673028
1         -4024.803754
2         -5623.422610
3         -4124.747207
4        340266.072047
             ...      
21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, Length: 21525, dtype: float64


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo.
</div>

In [166]:
# Comprueba el resultado - asegúrate de que esté arreglado

print(df['days_employed'].describe())
dfl4=df.loc[df['days_employed']<=0 ,:]

percentaje_days_employed=len(dfl4)/len(df)
print(f'\n porcentaje de valores negativos is {percentaje_days_employed:.2%}')

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

 porcentaje de valores negativos is 73.90%


En este caso usaremos como criterio para los valores grandes y muy atípicos el valor de 365*50=18250.

In [167]:
### correccion de valores Atipicos 

#### el criterio del margen es que el tiempo util de trabajo de una presona durante un año maximo es 50
margen_atipico_positivo=365*50
margen_atipico_negativo=-365*50


print(f'Margen de valor atipico inferior {margen_atipico_negativo} y valor atipico superior {margen_atipico_positivo}  ')
df_days_atipico_superior=df.query('days_employed>@margen_atipico_positivo')
df_days_atipico_inferior=df.query('days_employed<@margen_atipico_negativo')

days_atipico=len(df_days_atipico_superior)+len(df_days_atipico_inferior)
#df.loc[df['days_employed']<=0 ,:]

atipico_days_employed=days_atipico/len(df)
print(f'\n porcentaje de valores atipicos es {atipico_days_employed:.2%} antes de Corregir')


############### Correcion
def corregir_days_employed(days_employed):
    """
    La funcion multiplica por 0.01 si el parametro sobrepasa el margen
    """
    valor=days_employed
    if margen_atipico_negativo<=days_employed >= margen_atipico_positivo:
        valor=(valor/100)     
    return valor

df['days_employed'] = df['days_employed'].apply(corregir_days_employed)


################

df_days_atipico=df.query('@margen_atipico_negativo<days_employed>@margen_atipico_positivo')
atipico_days_employed=len(df_days_atipico)/len(df)
print(f'\n porcentaje de valores atipicos es {atipico_days_employed:.2%} despues de Corregir')

Margen de valor atipico inferior -18250 y valor atipico superior 18250  

 porcentaje de valores atipicos es 16.01% antes de Corregir

 porcentaje de valores atipicos es 0.00% despues de Corregir


In [168]:
print(df['days_employed'].describe())

count    19351.000000
mean     -1284.309490
std       3105.769115
min     -18388.949901
25%      -2747.423625
50%      -1203.369529
75%       -291.095954
max       4017.554005
Name: days_employed, dtype: float64


Vamos a convertir los valores negativos a positivos.

In [169]:
df['days_employed']=abs(df['days_employed'])



In [170]:
print(df['days_employed'].describe())

count    19351.000000
mean      2583.921756
std       2149.032624
min         24.141633
25%        927.009265
50%       2194.220567
75%       3658.750303
max      18388.949901
Name: days_employed, dtype: float64


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Esta columna tiene tres problemas:
    
1. Datos ausentes
2. Datos negativos
3. Datos anormalmente grandes

El primer problema lo abordarás más adelante, el segundo ya lo resolviste pero aún es necesario trabajar en el tercero.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante</b>  <a class="tocSkip"></a> <br>
    Listo. Se corrigio los posibles valores atipicos de la columna 'days_employed' , Se lo hizo agregando una celda.
</div>


<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Todos los datos positivos, antes de la transformación de negativos a positivos son inconsistentes. Así que antes de aplicar valor absoluto, puedes hacer el tratamiento que planteas anteriormente.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante #2 </b>  <a class="tocSkip"></a> <br>
    Listo. Se corrigio los posibles valores atipicos de la columna 'days_employed' , con el criterio de mayores de 50 años serian atipicos y su valor se reemplazara por su mismo valor dividido para 10. Se aplico el filtro de valores atipicos y despues se los convirtio en positivos como estaba aconsejado.
</div>

__Columna `dob_years`__ 

Revisando la Columna `dob_years` para encontrar posibles valores sospechosos.

In [171]:
# Revisa `dob_years` en busca de valores sospechosos y cuenta el porcentaje

df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

**Revisando columna `dob_years`**

+ Estos valores de deben convertir en enteros para ocupar menos espacio en memoria.
+ Podemos ver un mínimo de 0 años lo que lo hace algo inusual.
+ Notando se ve que los Años 0 son 100 datos.


In [172]:
# Resuelve los problemas en la columna `dob_years`, si existen


## contar los anyos donde es 0 
df_anyos_inusual=df.query('dob_years==0')
print('dob_year inusual:',len(df_anyos_inusual))

### Tratar valoes de 0, se sustituitan aquellos valores con un promedio de la edad basandose en la columna mas representativa
### en este caso se utilizara la columna income_type

df_retiree=df.query('income_type=="retiree"').query('dob_years !=0')
df_business=df.query('income_type=="business"').query('dob_years !=0')
df_employee=df.query('income_type=="employee"').query('dob_years !=0')
df_civil_servant=df.query('income_type=="civil servant"').query('dob_years !=0')

##### reemplazamos los valores por la media de la edad
def corregir_anios_0(row):
    edad = row['dob_years']
    income_type = row['income_type']
    retorno_edad=edad
    if edad ==0:
        if income_type=='retiree':
            retorno_edad=df_retiree['dob_years'].median()
            return round(retorno_edad)
        if income_type=='business':
            retorno_edad=df_business['dob_years'].median()
            return round(retorno_edad)
        if income_type=='employee':
            retorno_edad=df_employee['dob_years'].median()
            return round(retorno_edad) 
        if income_type=='civil servant':
            retorno_edad=df_civil_servant['dob_years'].median()
            return round(retorno_edad) 
    return edad
            

df['dob_years']=df.apply(corregir_anios_0,axis=1)

###cambiar el tipo




dob_year inusual: 101


In [173]:
# Comprueba el resultado - asegúrate de que esté arreglado
df_anyos_inusual=df.query('dob_years==0')
print('dob_year inusual:',len(df_anyos_inusual))


df['dob_years'].describe()


dob_year inusual: 0


count    21525.000000
mean        43.496167
std         12.231538
min         19.000000
25%         34.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Dado que es un dataset de individuos con historial crediticio, el valor de 0 para los años es inusual y debe ser corregido. 
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante</b>  <a class="tocSkip"></a> <br>
    Listo. Se trato los valores donde `dob_year`=0 , Se lo reeemplazo con la media de su equivalente en `income_type`
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo, pero cambia el estadístico `mean` por `median` será mejor dado que nuestros datos presentan una leve asimetría y además son discretos.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante #2</b>  <a class="tocSkip"></a> 
    
Listo. Se trato los cambios el `mean()` por `median()`
</div>

__Revisando columna `family_status`__:
+ El único error posible puede ser que no estén con el mismo formato de texto.
+ También pude ser que haya valores que sean numéricos.


In [174]:
# Veamos los valores de la columna
print(df['family_status'].describe())

df['family_status'].unique()


count       21525
unique          5
top       married
freq        12380
Name: family_status, dtype: object


array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In [175]:
df['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

+ Podemos apreciar que no hay valores raros dentro de la columna.

In [176]:
# Aborda los valores problemáticos en `family_status`, si existen



##tenemos un valor string numerico , lo debemos reemplazar por su equivalente.  
def reemplazar_family_status(row):
    """
    funcion que cambia los valores enteros por su equivalente string
    """

    family_status = row['family_status']
    family_status_id = row['family_status_id']

    if (isinstance(family_status,int)):
        if family_status_id==0:
            family_status='married'
        if family_status_id==1:
            family_status='civil partnership'
        if family_status_id==2:
            family_status='widow / widower'
        if family_status_id==3:
            family_status='divorced'
        if family_status_id==4:
            family_status='unmarried'
            
    return family_status
    


#df['family_status'] = df.apply(reemplazar_family_status, axis=1)




In [177]:
# Comprueba el resultado - asegúrate de que esté arreglado


### podemos comprobar que eliminamos el numero de esa columna

print(df['family_status'].describe())

df['family_status'].unique()


count       21525
unique          5
top       married
freq        12380
Name: family_status, dtype: object


array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

+ Tenemos que ya no hay valores enteros dentro de la columna por loque ya se ha corregido.

__Revisando columna `gender`:__

+ Los posible valores erróneos de la columna `gender` son que existan un valores numéricos.


In [178]:
# Veamos los valores en la columna

print(df['gender'].describe())


df['gender'].unique()

count     21525
unique        3
top           F
freq      14236
Name: gender, dtype: object


array(['F', 'M', 'XNA'], dtype=object)

+ Podemos apreciar que no hay valores raros, Están la 3 categorías esperadas.

In [179]:
# Aborda los valores problemáticos, si existen


###se reemplazara con 'xna' los valores numericos.
def reemplazar_gender(row):

    gender= row['gender']
    
    
    
    if (isinstance(gender,int)):
        gender='XNA'
        
    return gender
    
    
    
#df['gender'] = df.apply(reemplazar_gender, axis=1)
    

In [180]:
# Comprueba el resultado - asegúrate de que esté arreglado

print(df['gender'].describe())


df['gender'].unique()


count     21525
unique        3
top           F
freq      14236
Name: gender, dtype: object


array(['F', 'M', 'XNA'], dtype=object)

__Revisando la columna `income_type`:__

+ Vamos a revisar si hay valores raros dentro de la columna.



In [181]:
# Veamos los valores en la columna

print(df['income_type'].describe())


df['income_type'].unique()

count        21525
unique           8
top       employee
freq         11119
Name: income_type, dtype: object


array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [182]:
# Aborda los valores problemáticos, si existen


#### vemos que sigue habiendo el mismo valor numerico

def reemplazar_income_type(row):

    income_type= row['income_type']
    
    if income_type==0.5432752613240418:
        income_type='unknow'
        
    return income_type
    
    
    
#df['income_type'] = df.apply(reemplazar_income_type, axis=1)

In [183]:
# Comprueba el resultado - asegúrate de que esté arreglado

print(df['income_type'].describe())


df['income_type'].unique()

count        21525
unique           8
top       employee
freq         11119
Name: income_type, dtype: object


array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

### Comprobar valores duplicados del set de Datos

In [184]:
# Comprobar los duplicados

print(df.duplicated().sum())

print(len(df))
dfz1=len(df) ## filas antes de eliminar duplicados

71
21525


In [185]:
# Aborda los duplicados, si existen

df=df.drop_duplicates().reset_index(drop=True)

In [186]:
# Última comprobación para ver si tenemos duplicados


print(df.duplicated().sum())

0


In [187]:
# Comprueba el tamaño del conjunto de datos que tienes ahora, después de haber ejecutado estas primeras manipulaciones

print(len(df))
dfzz1=len(df)/dfz1 ###proporcion de eliminar duplicados (ahora , antes)
print(f'el cambio de porcentaje coservado despues de eliminar los duplicados fue {dfzz1:.2%}')

21454
el cambio de porcentaje coservado despues de eliminar los duplicados fue 99.67%


**Respuesta `Estudiante`**


Lo que se hizo con los datos fue:

   + Analizar valores problemáticos y reemplazarlos por valores representativos.
   + Cambiar los valores de 0 por valores representativos de la columna que representativa.
   + Eliminar los duplicados de las columnas .




<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo con los duplicados.
</div>

- [__Click para ir a la Tabla de Contenidos__](#tablacontenido)

In [188]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21454 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## 03. Trabajar con valores ausentes <a class="anchor" id="ausentes"></a>

**Trabajar con valores ausentes:**

+ Se va a crear un diccionario usando IDs , basándose en el siguiente formato: 

In [189]:
# Encuentra los diccionarios

#### se debe crear una columna de IDs
#### 


##'000000F' + (df.index + 1)+'-x'

df['id'] = (df.index+1).astype(str)+df['family_status_id'].astype(str) + df['education_id'].astype(str)

print(df["id"].describe())

######
"""
education_dict = dict(zip(df['education'], df['education_id']))
print(education_dict)

family_status_dict = dict(zip(df['family_status'], df['family_status_id']))
print(family_status_dict)
"""


count     21454
unique    21454
top         100
freq          1
Name: id, dtype: object


"\neducation_dict = dict(zip(df['education'], df['education_id']))\nprint(education_dict)\n\nfamily_status_dict = dict(zip(df['family_status'], df['family_status_id']))\nprint(family_status_dict)\n"

Crear una columna `id` nos ayudara a analizar los datos más adelante  

<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Debes crear un diccionario asociando `education` con `education_id` y otro para asociar `family_status` con `family_status_id`.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante #2</b>  <a class="tocSkip"></a> <br>
    Listo. se creo una columna `id` que asocia el `education_id` con `family_status`
    
</div>

In [190]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21454 non-null  object 
 12  id                21454 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


### Restaurar valores ausentes en `total_income`


Vamos a clasificar por edad creando una nueva columna.
Donde las personas menores de 18 serán consideradas como `children`, quienes tengan entre 18 y menor de 64 serán `adult` y mayores `retired`. 

Usaremos esta función para reemplazar los valores ausentes por un valor representativo. en este caso la nueva columna que vamos creando.


In [191]:
# Vamos a escribir una función que calcule la categoría de edad


def age_group(age):

    if age <= 18:
        return 'children'
    if age <= 64:
        return 'adult'
    return 'retired' 

In [192]:
# Prueba si la función funciona bien
print(age_group(4))
print(age_group(30))
print(age_group(65))

children
adult
retired


In [193]:
# Crear una nueva columna basada en la función

df['age_category']=df['dob_years'].apply(age_group)

In [194]:
# Comprobar cómo los valores en la nueva columna

df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42    bachelor's degree             0   
1          1    4024.803754         36  secondary education             1   
2          0    5623.422610         33  secondary education             1   
3          3    4124.747207         32  secondary education             1   
4          0    3402.660720         53  secondary education             1   
5          0     926.185831         27    bachelor's degree             0   
6          0    2879.202052         43    bachelor's degree             0   
7          0     152.779569         50  secondary education             1   
8          2    6929.865299         35    bachelor's degree             0   
9          0    2188.756445         41  secondary education             1   
10         2    4171.483647         36    bachelor's degree             0   
11         0     792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0    1846.641941         54         some college             2   
14         0    1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose    id age_category  
0      40620.102                purchase of the house   100        adult  
1      17932.802                         car purchase   201        adult  
2      23341.752                purchase of the house   301        adult  
3      42820.568              supplementary education   401        adult  
4      25378.572                    to have a wedding   511        adult  
5      40922.170                purchase of the house   610        adult  
6      38484.156                 housing transactions   700        adult  
7      21731.829                            education   801        adult  
8      15337.093                     having a wedding   910        adult  
9      23108.150  purchase of the house for my family  1001        adult  
10     18230.959                      buy real estate  1100        adult  
11     12331.077           buy commercial real estate  1201        adult  
12           NaN                    to have a wedding  1311      retired  
13     20873.317                         car purchase  1402        adult  
14     26420.466          buy residential real estate  1510        adult

Vamos a crear un data frame seleccionando los valores que no sean nulos, y de aquí usaremos los valores para remplazarlos en las filas del data frame original donde haya datos vacíos.

In [195]:
# Crea una tabla sin valores ausentes y muestra algunas de sus filas para asegurarte de que se ve bien


### Restaurando valores Ausentes de 'total_income'

print('\nColumna de valores del Dataset :\n',df['total_income'].head(15))


##### tabla filtrada con valores no ausentes
dfl5=df.loc[df['total_income'].notnull(),:]

print('\nColumna de valores no ausentes:\n',dfl5['total_income'].head(15))


Columna de valores del Dataset :
 0     40620.102
1     17932.802
2     23341.752
3     42820.568
4     25378.572
5     40922.170
6     38484.156
7     21731.829
8     15337.093
9     23108.150
10    18230.959
11    12331.077
12          NaN
13    20873.317
14    26420.466
Name: total_income, dtype: float64

Columna de valores no ausentes:
 0     40620.102
1     17932.802
2     23341.752
3     42820.568
4     25378.572
5     40922.170
6     38484.156
7     21731.829
8     15337.093
9     23108.150
10    18230.959
11    12331.077
13    20873.317
14    26420.466
15    18691.345
Name: total_income, dtype: float64


In [196]:
df.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    3402.660720         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose   id age_category  
0    purchase of the house  100        adult  
1             car purchase  201        adult  
2    purchase of the house  301        adult  
3  supplementary education  401        adult  
4        to have a wedding  511        adult

In [197]:
# Examina los valores medios de los ingresos en función de los factores que identificaste

##valor representativo media:
valor_representativo_total_income_mean=dfl5['total_income'].mean()
print(valor_representativo_total_income_mean)

26787.56835465871


In [198]:
# Examina los valores medianos de los ingresos en función de los factores que identificaste


##valor reresentativo mediana
valor_representativo_total_income_median=dfl5['total_income'].median()
print(valor_representativo_total_income_median)

23202.87


Se reemplazara las filas vacías por la media de los datos no vacíos que pertenecen a su respectiva categoría.En este caso por `income_type`.

In [199]:
df_total_income_retiree=df.query('income_type=="retiree"').query('income_type.notnull()')
df_total_income_business=df.query('income_type=="business"').query('income_type.notnull()')
df_total_income_employee=df.query('income_type=="employee"').query('income_type.notnull()')
df_total_income_civil_servant=df.query('income_type=="civil servant"').query('income_type.notnull()')

In [200]:
#  Escribe una función que usaremos para completar los valores ausentes

##estadistico General
#df['total_income']=df['total_income'].fillna(valor_representativo_total_income_median)
#### la mediana es un valor mas representattivo.



##Se Utilizara una Funcion en base a `income_type`

def llenando_vacios_de_total_income(row):
    """
    #funcion que devuelve donde es vacio`total_income` lo reemplaza por un representativo de `income_type`
    """
    total_income = row['total_income']
    income_type = row['income_type']
    retorno=total_income
    
    if math.isnan(total_income):
        if income_type=="retiree":
            retorno=df_total_income_retiree['total_income'].median()
            return retorno
        if income_type=="business":
            retorno=df_total_income_business['total_income'].median()
            return retorno
        if income_type=="employee":
            retorno=df_total_income_employee['total_income'].median()
            return retorno
        if income_type=="civil servant":
            retorno=df_total_income_civil_servant['total_income'].median()
            return retorno
        else:
            retorno =valor_representativo_total_income_median
            return retorno

    return retorno




<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo.
</div>

In [201]:
# Comprueba si funciona
print(df['total_income'].head(15))

#### antes era NaN
### mirando el indice 12 se verifica que se sustituyo.
print(df.loc[12,'total_income'],'\n')


0     40620.102
1     17932.802
2     23341.752
3     42820.568
4     25378.572
5     40922.170
6     38484.156
7     21731.829
8     15337.093
9     23108.150
10    18230.959
11    12331.077
12          NaN
13    20873.317
14    26420.466
Name: total_income, dtype: float64
nan 



In [202]:
# Aplícalo a cada fila


df['total_income']=df.apply(llenando_vacios_de_total_income,axis=1)



In [203]:
# Comprueba si tenemos algún error
df.head(15)

### mirando el indice 12 se verifica que se sustituyo.
print('ANTES ERA VALOR NULO: ',df.loc[12,'total_income'],'\n')

ANTES ERA VALOR NULO:  18962.318 



En este caso botaba un valor nulo dicha columna , por lo que se agregó un else a la función `llenando_vacios_de_total_income` para que tome ese valor faltante.

In [204]:
# Reemplazar los valores ausentes si hay algún error


####no ha habido error en esta columna.
print(df['total_income'].notnull().sum())
len(df)


21454


21454

Ahora sin lugar a duda se trato la Columna `total_income` de la manera correcta.

In [205]:
# Comprobar el número de entradas en las columnas

df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  id                21454 non-null  object 
 13  age_category      21454 non-null  object 
dtypes: float64(2), int64(5), object(7)
memory usage: 2.3+ MB


###  Restaurar valores en `days_employed` <a class="anchor" id="restaurar"></a>

Se usara el mismo criterio que la tabla anterior , se reemplazara por una media representativa , en este caso en funcion de una columna que no tenga datos vacios como `income_type`.

In [206]:
# Distribución de las medianas de `days_employed` en función de los parámetros identificados


df['days_employed'].median()

2194.220566878695

In [207]:
# Distribución de las medias de `days_employed` en función de los parámetros identificados


print("promedio days_employed",df['days_employed'].mean())

df['days_employed'].describe()

promedio days_employed 2583.921755515992


count    19351.000000
mean      2583.921756
std       2149.032624
min         24.141633
25%        927.009265
50%       2194.220567
75%       3658.750303
max      18388.949901
Name: days_employed, dtype: float64


Para reemplazar los valores vamos a sacar un promedio de su categoría y reemplazarlos. La media de su respectiva categoría representa el valor más similar.


In [208]:
df_days_employed_retiree=df.query('income_type=="retiree"').query('days_employed.notnull()')
df_days_employed_business=df.query('income_type=="business"').query('days_employed.notnull()')
df_days_employed_employee=df.query('income_type=="employee"').query('days_employed.notnull()')
df_days_employed_civil_servant=df.query('income_type=="civil servant"').query('days_employed.notnull()')

In [209]:
# Escribamos una función que calcule medias o medianas (dependiendo de tu decisión) según el parámetro identificado

## esto es un nuevo df que calculara medias o medianas para reemplazar a los valores ausentes.
dfl6=df.loc[df['days_employed'].notnull(),:]
mediana_dfl6=dfl6['days_employed'].median()

#print(mediana_dfl6)


###funcion 

def llenando_vacios_de_days_employed(row):
    """
    #funcion que devuelve donde es vacio`days_employed` lo reemplaza por un representativo de `income_type`
    """
    days_employed = row['days_employed']
    income_type = row['income_type']
    retorno=days_employed
    
    if math.isnan(days_employed):
        if income_type=="retiree":
            retorno=df_days_employed_retiree['days_employed'].median()
            return retorno
        if income_type=="business":
            retorno=df_days_employed_business['days_employed'].median()
            return retorno
        if income_type=="employee":
            retorno=df_days_employed_employee['days_employed'].median()
            return retorno
        if income_type=="civil servant":
            retorno=df_days_employed_civil_servant['days_employed'].median()
            return retorno
        else:
            retorno =mediana_dfl6
            return retorno

    return retorno


La función valdrá con un data frame. No con una instancia así que se le probara directamente en el data frame.

In [210]:
# Comprueba que la función funciona
##print(df['days_employed'].head(15))
##dfl6['days_employed'].head(15)

row_probar_funcion_llenar_days_employed = {'days_employed': np.nan, 'income_type': 'business'}

print(row_probar_funcion_llenar_days_employed)

print(df_days_employed_business['days_employed'].mean())
llenando_vacios_de_days_employed(row_probar_funcion_llenar_days_employed)
print(row_probar_funcion_llenar_days_employed)


{'days_employed': nan, 'income_type': 'business'}
2111.5243982977295
{'days_employed': nan, 'income_type': 'business'}


Se sabe que para probar la función se debe usar un dato del tipo DataFrame , por lo que mejor será aplicarlo directo a la función.

In [211]:
# Aplicar la función al income_type

print(df['days_employed'].isna().sum())

#df['days_employed']=df['days_employed'].fillna(mediana_dfl6)


2103


In [212]:
# Comprueba si la función funcionó

df['days_employed'].isna().sum()

2103

In [213]:
# Reemplazar valores ausentes

df['days_employed']=df.apply(llenando_vacios_de_days_employed,axis=1)

df['days_employed'].isna().sum()

### ya se ha reemplazado arriba

0

Ahora podemos Observar que no existe ningún valor nulo dentro de nuestros datos.

In [214]:
# Comprueba las entradas en todas las columnas: asegúrate de que hayamos corregido todos los valores ausentes

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  id                21454 non-null  object 
 13  age_category      21454 non-null  object 
dtypes: float64(2), int64(5), object(7)
memory usage: 2.3+ MB


<div class="alert alert-block alert-danger">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo con esta variable pero habrá que ejecutarla de nuevo una vez se corrija la sección más arriba.
    
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante</b>  <a class="tocSkip"></a> <br>
    Listo. Se corrigio la seccion de arriba.
    
</div>

- [__Click para ir a la Tabla de Contenidos__](#tablacontenido)

## 04 Clasificación de datos <a class="anchor" id="clasificar"></a>

Se puede apreciar que en la columna `purpose` es algo difícil de analizar, pero hay palabras que se pueden repetir.
Así que vamos a agrupar aplicando un algoritmo para que sea más sencillo.
En este caso lo agruparemos una función que se verá más adelante.


In [215]:
# Muestra los valores de los datos seleccionados para la clasificación

df['purpose'].describe()



count                21454
unique                  38
top       wedding ceremony
freq                   791
Name: purpose, dtype: object

Comprobar los valores únicos:

In [216]:
# Comprobar los valores únicos

print(df['purpose'].unique())

array_unique=df['purpose'].unique()


['purchase of the house' 'car purchase' 'supplementary education'
 'to have a wedding' 'housing transactions' 'education' 'having a wedding'
 'purchase of the house for my family' 'buy real estate'
 'buy commercial real estate' 'buy residential real estate'
 'construction of own property' 'property' 'building a property'
 'buying a second-hand car' 'buying my own car'
 'transactions with commercial real estate' 'building a real estate'
 'housing' 'transactions with my real estate' 'cars' 'to become educated'
 'second-hand car purchase' 'getting an education' 'car'
 'wedding ceremony' 'to get a supplementary education'
 'purchase of my own house' 'real estate transactions'
 'getting higher education' 'to own a car' 'purchase of a car'
 'profile education' 'university education'
 'buying property for renting out' 'to buy a car' 'housing renovation'
 'going to university']


La función "change_purpose" es una función de clasificación de datos que se utiliza para asignar una categoría específica a los datos de propósito de un préstamo. El objetivo de esta función es ayudar a responder a las preguntas planteadas y probar las diferentes hipótesis en el proyecto de análisis de datos. La función toma como entrada el propósito de un préstamo y devuelve una categoría específica.

La función utiliza una estructura de control de flujo "if-elif" para determinar la categoría correcta para el propósito del préstamo. Por ejemplo, si el propósito incluye la palabra "wedding", la función asignará la categoría "wedding". De lo contrario, si el propósito incluye cualquiera de las palabras "real", "state", "house", "property" o "housing", la función asignará la categoría "realestate". Esta función esta diseñada para clasificar los datos de texto, pero también puedes usar otras instrucciones para clasificar los datos numéricos. Es importante proporcionar una explicación clara de las decisiones tomadas en el proceso de clasificación de datos.






In [217]:
# Escribamos una función para clasificar los datos en función de temas comunes


def change_purpose(purpose):
    purpose_category='unknow'
    
    if 'wedding' in purpose:
        purpose_category='wedding'
       
    
    elif ('real' in purpose) or ('state' in purpose) or ('house' in purpose) or ('property' in purpose) or ('housing' in purpose):
        purpose_category='real estate'
       
    
    elif 'car' in purpose:
        purpose_category='car'
        
    elif ('education' in purpose) or ('educated' in purpose) or ('university' in purpose):
        purpose_category='education'
    return purpose_category

In [218]:
# Crea una columna con las categorías y cuenta los valores en ellas
df['purpose']=df['purpose'].apply(str)

df['purpose_category'] = df['purpose'].apply(change_purpose)

df['purpose_category'].value_counts()


real estate    10811
car             4306
education       4013
wedding         2324
Name: purpose_category, dtype: int64

En este caso de tantos valores que había en categoría , al aplicar la función se lo simplifico quedando solamente 5 categorías:
`unknow`, `education`, `wedding`, `car`, `real_state`.


In [219]:
# Revisar todos los datos numéricos en la columna seleccionada para la clasificación

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  id                21454 non-null  object 
 13  age_category      21454 non-null  object 
 14  purpose_category  21454 non-null  object 
dtypes: float64(2), int64(5), object(8)
memory usage: 2.5+ MB


In [220]:
# Obtener estadísticas resumidas para la columna

df['total_income'].describe()

count     21454.000000
mean      26448.571779
std       15705.750220
min        3306.762000
25%       17219.817250
50%       22815.103500
75%       31330.237250
max      362496.645000
Name: total_income, dtype: float64

La función "`clasificar_total_income()`" es una función que permite clasificar los datos de ingresos totales de los clientes en diferentes grupos numéricos basándose en rangos.

La función utiliza las librería de numpy para calcular los quantiles con saltos del 20% de los datos de ingreso total (q1,q2,q3,q4). Estos quantiles serán utilizados como límites para clasificar los datos en diferentes categorías.

La función toma como entrada el ingreso total de un cliente y utiliza una estructura de control de flujo "if-elif" para determinar en qué categoría debe clasificar el ingreso. Por ejemplo, si el ingreso de un cliente es menor o igual al primer quantil (q1), la función asignará la categoría "0 very low", si el ingreso esta entre el primer y segundo quantil la función asignara la categoria "1 low" y así sucesivamente. Finalmente, la función devuelve la categoría asignada.

In [221]:
# Crear una función para clasificar en diferentes grupos numéricos basándose en rangos


#### en este caso clasificaresmos el total income
#####usaremos quantiles con saltos del 20%  para tener categorias



q1=np.quantile(df['total_income'],0.2)
q2=np.quantile(df['total_income'],0.4)
q3=np.quantile(df['total_income'],0.6)
q4=np.quantile(df['total_income'],0.8)

def clasificar_total_income(income):
    categoria_income=''
    if income<=q1:
        categoria_income='0 very low'
    elif (income>q1) and (income<=q2):
        categoria_income='1 low'
    elif (income>q2) and (income<=q3):
        categoria_income='2 medium'
    elif (income>q3) and (income<=q4):
        categoria_income='3 high'
    else:
        categoria_income='4 very high'
    
    return categoria_income

In [222]:
# Crear una columna con categorías
df['income_category']=df['total_income'].apply(clasificar_total_income)

In [223]:
# Contar los valores de cada categoría para ver la distribución

print(df['income_category'].value_counts())

####como se hizo por cuantiles se logro que tengan aproximadamente la misma cantidad de datos.

df['income_category'].describe()


4 very high    4291
1 low          4291
0 very low     4291
3 high         4291
2 medium       4290
Name: income_category, dtype: int64


count           21454
unique              5
top       4 very high
freq             4291
Name: income_category, dtype: object

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo en esta sección.
</div>

- [__Click para ir a la Tabla de Contenidos__](#tablacontenido)

## 05. Comprobación de las hipótesis <a class="anchor" id="comprobacion"></a>


**¿Existe una correlación entre tener hijos y pagar a tiempo?**

In [224]:
# Comprueba los datos sobre los hijos y los pagos puntuales
df_hijos_pago_0 = df.loc[df['debt']==0].groupby('debt')['children'].value_counts()

df_hijos_pago_1= df.loc[df['debt']==1].groupby('debt')['children'].value_counts()


# Calcular la tasa de incumplimiento en función del número de hijos

print(df_hijos_pago_0.head(6))

print(df_hijos_pago_1.head(6))






debt  children
0     0           13028
      1            4410
      2            1926
      3             303
      4              37
      5               9
Name: children, dtype: int64
debt  children
1     0           1063
      1            445
      2            202
      3             27
      4              4
Name: children, dtype: int64


In [225]:

pivot_table=pd.pivot_table(df, values='id', index='children', columns='debt', aggfunc=['count'], sort=True)
##rellenando valores vacios del pivot table paraque no altere.
pivot_table['count', 1].fillna(value=0, inplace=True)

pivot_table['tasa %',0] = pivot_table['count', 0] / (pivot_table['count', 0]+pivot_table['count', 1]) * 100
pivot_table['tasa %',1] = pivot_table['count', 1] / (pivot_table['count', 0]+pivot_table['count', 1]) * 100

pivot_table

count              tasa %          
debt            0       1           0         1
children                                       
0         13028.0  1063.0   92.456178  7.543822
1          4410.0   445.0   90.834192  9.165808
2          1926.0   202.0   90.507519  9.492481
3           303.0    27.0   91.818182  8.181818
4            37.0     4.0   90.243902  9.756098
5             9.0     0.0  100.000000  0.000000


**Respuesta `H1`:**

+ No se puede concluir que la falta de pago tiene relacion con los hijos que se se tiene , ya que tiene una proporcion similar, vemos que el porcentaje de la table de las personas que no tienen deuda se mantiene constante, ademas las personas que no tienen hijos son las que mejor pagan hasta el momento por una pequeña diferencia.



**¿Existe una correlación entre la situación familiar y el pago a tiempo?**

In [226]:
# Comprueba los datos del estado familiar y los pagos a tiempo

df_familia_status_pago_0 = df.loc[df['debt']==0].groupby('debt')['family_status'].value_counts()

df_familia_status_pago_1= df.loc[df['debt']==1].groupby('debt')['family_status'].value_counts()

# Calcular la tasa de incumplimiento basada en el estado familiar

print(df_familia_status_pago_0,"\n")

print(df_familia_status_pago_1)



debt  family_status    
0     married              11408
      civil partnership     3763
      unmarried             2536
      divorced              1110
      widow / widower        896
Name: family_status, dtype: int64 

debt  family_status    
1     married              931
      civil partnership    388
      unmarried            274
      divorced              85
      widow / widower       63
Name: family_status, dtype: int64


In [227]:
pivot_table_family=pd.pivot_table(df, values='id', index='family_status', columns='debt', aggfunc=['count'], sort=True)

pivot_table_family['tasa %',0] = pivot_table_family['count', 0] / (pivot_table_family['count', 0]+pivot_table_family['count', 1]) * 100
pivot_table_family['tasa %',1] = pivot_table_family['count', 1] / (pivot_table_family['count', 0]+pivot_table_family['count', 1]) * 100

pivot_table_family

count          tasa %          
debt                   0    1          0         1
family_status                                     
civil partnership   3763  388  90.652855  9.347145
divorced            1110   85  92.887029  7.112971
married            11408  931  92.454818  7.545182
unmarried           2536  274  90.249110  9.750890
widow / widower      896   63  93.430657  6.569343


**Respuesta `H2`:**
 + Podemos concluir que que el 92.45 % de las personas  se encuentran casadas no presentan deudas o las pagan a tiempo, pero tambien los divorciados y los viudos por lo que se debe procurar dar prestamos a ese grupo de personas porque las pagan a tiempo.
 
+ Puedo afirmar que existe una correlación entre la situación familiar y el pago a tiempo. Se observa que cada estado civil tiene una tasa de incumplimiento de pago diferente, siendo los clientes solteros quienes tienen una tasa de incumplimiento de pago más alta en comparación con los clientes casados o con pareja. 

**¿Existe una correlación entre el nivel de ingresos y el pago a tiempo?**

In [228]:
# Comprueba los datos del nivel de ingresos y los pagos a tiempo

df_total_income_pago_0 = df.loc[df['debt']==0].groupby('debt')['income_category'].value_counts()

df_total_income_pago_1= df.loc[df['debt']==1].groupby('debt')['income_category'].value_counts()


# Calcular la tasa de incumplimiento basada en el nivel de ingresos

print(df_total_income_pago_0,"\n")

print(df_total_income_pago_1)


debt  income_category
0     4 very high        3991
      0 very low         3947
      1 low              3930
      3 high             3930
      2 medium           3915
Name: income_category, dtype: int64 

debt  income_category
1     2 medium           375
      1 low              361
      3 high             361
      0 very low         344
      4 very high        300
Name: income_category, dtype: int64


In [229]:
pivot_table_income_category=pd.pivot_table(df, values='id', index='income_category', columns='debt', aggfunc=['count'], sort=True)
##rellenando valores vacios del pivot table paraque no altere.


pivot_table_income_category['tasa %',0] = pivot_table_income_category['count', 0] / (pivot_table_income_category['count', 0]+pivot_table_income_category['count', 1]) * 100
pivot_table_income_category['tasa %',1] = pivot_table_income_category['count', 1] / (pivot_table_income_category['count', 0]+pivot_table_income_category['count', 1]) * 100

pivot_table_income_category

count          tasa %          
debt                0    1          0         1
income_category                                
0 very low       3947  344  91.983221  8.016779
1 low            3930  361  91.587043  8.412957
2 medium         3915  375  91.258741  8.741259
3 high           3930  361  91.587043  8.412957
4 very high      3991  300  93.008623  6.991377

**Respuesta `H3`:**

+ Se concluye que las personas que han tenido mayor ingreso pueden pagar su deuda, o son mejor paga. Esto los vuelve candidatos para prestamos.
+ Sí, existe una correlación entre el nivel de ingresos y el pago a tiempo. Según los datos presentados en la tabla, se puede observar que a medida que aumenta el nivel de ingresos, también aumenta la tasa de pago a tiempo. Los clientes con un nivel de ingresos muy bajo tienen una tasa de pago a tiempo del 91.98%, mientras que los clientes con un nivel de ingresos muy alto tienen una tasa de pago a tiempo del 93.00%.


**¿Cómo afecta el propósito del crédito a la tasa de incumplimiento?**

In [230]:
pivot_table_purpose_category=pd.pivot_table(df, values='id', index='purpose_category', columns='debt', aggfunc=['count'], sort=True)
##rellenando valores vacios del pivot table paraque no altere.


pivot_table_purpose_category['tasa %',0] = pivot_table_purpose_category['count', 0] / (pivot_table_purpose_category['count', 0]+pivot_table_purpose_category['count', 1]) * 100
pivot_table_purpose_category['tasa %',1] = pivot_table_purpose_category['count', 1] / (pivot_table_purpose_category['count', 0]+pivot_table_purpose_category['count', 1]) * 100

pivot_table_purpose_category

count          tasa %          
debt                  0    1          0         1
purpose_category                                 
car                3903  403  90.640966  9.359034
education          3643  370  90.779965  9.220035
real estate       10029  782  92.766627  7.233373
wedding            2138  186  91.996558  8.003442

**Respuesta `H4`:**


Los datos presentados en la tabla muestran que diferentes categorías de propósitos tienen tasas de incumplimiento de pago diferentes. Por ejemplo, se observa que los clientes que solicitan un préstamo para comprar un automóvil tienen una tasa de incumplimiento de pago del 9.36%, mientras que los clientes que solicitan un préstamo para la educación tienen una tasa de incumplimiento de pago del 9.22%.


<div class="alert alert-block alert-warning">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Gran trabajo en esta sección, Gabriel. Habrá que ejecutarla de nuevo una vez se implementen los cambios sugeridos más arriba.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante</b>  <a class="tocSkip"></a> <br>
    Listo. Se corrigio la seccion de arriba.
    
</div>

# Conclusión general <a class="anchor" id="conclusiongeneral"></a>


A partir de los datos analizados y las hipótesis propuestas, se puede concluir que existen varios factores que pueden afectar la tasa de incumplimiento de pago de un préstamo. 

En primer lugar, se observa que la situación familiar de los clientes tiene un impacto en el pago a tiempo, con los clientes casados o con pareja presentando una tasa de incumplimiento de pago más baja en comparación con los clientes solteros. 

Además, el nivel de ingresos de los clientes también se relaciona con el pago a tiempo, con los clientes con ingresos más altos teniendo una tasa de pago a tiempo más alta. 

Por último, el propósito del crédito también puede afectar la tasa de incumplimiento de pago, con diferentes categorías de propósitos presentando tasas de incumplimiento de pago diferentes.

En resumen, el informe final debe reflejar que se deben tener en cuenta estos factores al momento de evaluar la solvencia crediticia de un cliente potencial y asignarle una puntuación de crédito. Es importante mencionar que no existe un factor determinante para el incumplimiento de pago y se deben considerar varios factores a la hora de evaluar el riesgo de un cliente.
    

In [231]:
pivot_table

count              tasa %          
debt            0       1           0         1
children                                       
0         13028.0  1063.0   92.456178  7.543822
1          4410.0   445.0   90.834192  9.165808
2          1926.0   202.0   90.507519  9.492481
3           303.0    27.0   91.818182  8.181818
4            37.0     4.0   90.243902  9.756098
5             9.0     0.0  100.000000  0.000000

In [232]:
pivot_table_family

count          tasa %          
debt                   0    1          0         1
family_status                                     
civil partnership   3763  388  90.652855  9.347145
divorced            1110   85  92.887029  7.112971
married            11408  931  92.454818  7.545182
unmarried           2536  274  90.249110  9.750890
widow / widower      896   63  93.430657  6.569343

In [233]:
pivot_table_income_category

count          tasa %          
debt                0    1          0         1
income_category                                
0 very low       3947  344  91.983221  8.016779
1 low            3930  361  91.587043  8.412957
2 medium         3915  375  91.258741  8.741259
3 high           3930  361  91.587043  8.412957
4 very high      3991  300  93.008623  6.991377

In [234]:
pivot_table_purpose_category

count          tasa %          
debt                  0    1          0         1
purpose_category                                 
car                3903  403  90.640966  9.359034
education          3643  370  90.779965  9.220035
real estate       10029  782  92.766627  7.233373
wedding            2138  186  91.996558  8.003442

- [__Click para ir a la Tabla de Contenidos__](#tablacontenido)

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Esto debe hacerse con python exclusivamente para cumplir con los objetivos del curso. Puedes hacerlo con una tabla pivot.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante</b>  <a class="tocSkip"></a> <br>
    Listo. Se elimino la tabla excel que estaba y se reemplazo por su equivalente en pyhton.

</div>

<div class="alert alert-block alert-danger">

# Comentarios generales #1
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Hola, Gabriel. Has hecho un buen trabajo con el análisis de este dataset y su preprocesamiento, no obstante, es necesario trabajar en algunos puntos y el proyecto será aprobado cuando los mismos sean corregidos. He dejado comentarios detallados a lo largo del documento pero los enuemraré a continuación:
    
1. Elimina los comentarios entre corchetes y sustituyelos por conclusiones intermedias. El proyecto no se puede aprobar si esos comentarios están allí.
2. Revisa la sección de preprocesamiento. Una línea de código ha trastocado la estructura interna del dataset.
3. Solo podemos utilizar python para encontrar las soluciones a los problemas planteados en el proyecto.
4. La imputación de datos ausentes es mejor hacerla a través de estadísticos condicionados en lugar de estadísticos generales.
    
Una vez esto sea corregido, el proyecto será aprobado.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante</b>  <a class="tocSkip"></a> <br>
    Listo. Se corrigio lo mas los comentarios en rojo de la primera revision.

</div>

<div class="alert alert-block alert-danger">

# Comentarios generales #2
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Gran trabajo, Gabriel. Has corregido gran parte de lo sugerido pero aún nos quedan algunos puntos por solucionar. Te he dejado comentarios detallados a lo largo del documento para ello.
</div>

<div class="alert alert-block alert-info">
    
<b>Respuesta del estudiante #2</b>  <a class="tocSkip"></a> <br>
    Listo. Se hizo las correciones comentadas a detalle. 
    
</div>